### PR TITLE
feat: implement create and alter udf server

### PR DESCRIPTION
**Pull request by gitccl**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Extending the syntax and functionality of the `CREATE FUNCTION` and `ALTER FUNCTION` to support creating and altering a UDF Server in Databend.

Part of #12177

### CODE REVIEW

**Review by drmingdrmer — COMMENTED** [review]
I see that UDF is not used anywhere in our codebase, thus these incompatible changes will not break anything. That's good.

But for future compatibility, it is recommended to serialize these UDF types in protobuf, as TableMeta does(unless you are gonna be like to maintain compatibility manually):

https://github.com/datafuselabs/databend/blob/974f722746951bad3db9f0d794cffefa3ad150aa/src/meta/protos/proto/table.proto#L63-L66

And implement `FromToProto` for UDF types. Please refer to https://github.com/drmingdrmer/databend/blob/974f722746951bad3db9f0d794cffefa3ad150aa/src/meta/README.md#how-to-add-new-meta-data-types-to-store-in-meta-service

**Comment by gitccl** [comment]
> I see that UDF is not used anywhere in our codebase, thus these incompatible changes will not break anything. That's good.
> 
> But for future compatibility, it is recommended to serialize these UDF types in protobuf, as TableMeta does(unless you are gonna be like to maintain compatibility manually):
> 
> https://github.com/datafuselabs/databend/blob/974f722746951bad3db9f0d794cffefa3ad150aa/src/meta/protos/proto/table.proto#L63-L66
> 
> And implement `FromToProto` for UDF types. Please refer to https://github.com/drmingdrmer/databend/blob/974f722746951bad3db9f0d794cffefa3ad150aa/src/meta/README.md#how-to-add-new-meta-data-types-to-store-in-meta-service

Got it. I will use protobuf to serialize these UDF types.

**Review by drmingdrmer — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

File: src/query/ast/src/parser/statement.rs [hunk 1]
```diff
@@ -1506,7 +1488,7 @@ pub fn statement(i: Input) -> IResult<StatementMsg> {
             | #show_roles : "`SHOW ROLES`"
             | #create_role : "`CREATE ROLE [IF NOT EXISTS] '<role_name>']`"
             | #drop_role : "`DROP ROLE [IF EXISTS] '<role_name>'`"
-            | #create_udf : "`CREATE FUNCTION [IF NOT EXISTS] <udf_name> (<parameter>, ...) -> <definition expr> [DESC = <description>]`"
+            | #create_udf : "`CREATE FUNCTION [IF NOT EXISTS] <udf_name> AS {(<parameter>, ...) -> <definition expr> | (<arg_type>, ...) -> <return_type> ADDRESS <udf_server_address>} [DESC = <description>]`"
```
Comment (Xuanwo) [inline src/query/ast/src/parser/statement.rs:1491]: How about using `ADDRESS=<udf_server_address>` to align with other options?

Comment (gitccl) [reply src/query/ast/src/parser/statement.rs:1491]: Seems good.

**Review by xudong963 — APPROVED** [review]
LGTM, you can open a tracking issue to record your following work!

**Comment by Xuanwo** [comment]
Hi, I have changed the base branch to `udf` so that we can evaluate it first without blocking your continued development.

File: src/query/sql/src/planner/binder/udf.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                let mut arg_datatypes = Vec::with_capacity(arg_types.len());
+                for arg_type in arg_types {
+                    arg_datatypes.push(DataType::from(&resolve_type_name(arg_type)?));
+                }
+
+                Ok(UserDefinedFunction {
+                    name: udf_name.to_string(),
+                    description: udf_description.clone().unwrap_or_default(),
+                    definition: PlanUDFDefinition::UDFServer(UDFServer {
+                        address: address.clone(),
```
Comment (b41sh) [inline src/query/sql/src/planner/binder/udf.rs:75]: Is it necessary to check that address is a valid HTTP address?

Comment (gitccl) [reply src/query/sql/src/planner/binder/udf.rs:75]: Yes, I will do it in subsequent pr.

**Review by b41sh — APPROVED** [review]

**Pull request merged**